src/constraint/spec.ts

```diff
@@ -429,37 +430,64 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
   {
     name: 'omitNonPositionalOrFacetOverPositionalChannels',
     description: 'Do not use non-positional channels unless all positional channels are used',
-    properties: [Property.CHANNEL],
+    properties: [Property.CHANNEL, Property.MARK],
     allowWildcardForProperties: false,
     strict: false,
     satisfy: (specM: SpecQueryModel, _: Schema, opt: QueryConfig) => {
-      const encodings = specM.specQuery.encodings;
-      let hasNonPositionalChannelOrFacet = false;
-      let hasEnumeratedNonPositionOrFacetChannel = false;
-      let hasX = false, hasY = false;
-      for (let i = 0; i < encodings.length; i++) {
-        const encQ = encodings[i];
-        if (isValueQuery(encQ) || (isDisabledAutoCountQuery(encQ))) {
-          continue; // ignore skipped encoding
-        }
+      const mark = specM.getMark();
 
-        const channel = encQ.channel;
-        if (channel === Channel.X) {
-          hasX = true;
-        } else if (channel === Channel.Y) {
-          hasY = true;
-        } else if (!isWildcard(channel)) {
-          // All non positional channel / Facet
-          hasNonPositionalChannelOrFacet = true;
-          if (specM.wildcardIndex.hasEncodingProperty(i, Property.CHANNEL)) {
-            hasEnumeratedNonPositionOrFacetChannel = true;
+      switch(mark) {
+        case Mark.RECT:
+          // RECT is a special exception, as it can be used in a 1xD manner
+          // (e.g., linear heatmap).
+          return true;
+        default:
+          const encodings = specM.specQuery.encodings;
+          let hasNonPositionalChannelOrFacet = false;
+          let hasEnumeratedNonPositionOrFacetChannel = false;
+          let hasX = false, hasY = false;
+          for (let i = 0; i < encodings.length; i++) {
+            const encQ = encodings[i];
+            if (isValueQuery(encQ) || (isDisabledAutoCountQuery(encQ))) {
+              continue; // ignore skipped encoding
+            }
+
+            const channel = encQ.channel;
+            if (channel === Channel.X) {
+              hasX = true;
+            } else if (channel === Channel.Y) {
+              hasY = true;
+            } else if (!isWildcard(channel)) {
+              // All non positional channel / Facet
+              hasNonPositionalChannelOrFacet = true;
+              if (specM.wildcardIndex.hasEncodingProperty(i, Property.CHANNEL)) {
+                hasEnumeratedNonPositionOrFacetChannel = true;
+              }
+            }
           }
-        }
+
+          if ( hasEnumeratedNonPositionOrFacetChannel ||
+              (opt.constraintManuallySpecifiedValue && hasNonPositionalChannelOrFacet)
+            ) {
+            return hasX && hasY;
+          }
+          return true;
       }
+    }
+  },
+  {
+    name: 'omitFacetOverPositionalChannels',
```

Comment:
I still do not understand why you need to split this constraint from the other.  Could you please explain more?

Reply:
The split was motivated by 1D rect support. The original constraint also meant 'color' could not be used unless both positional channels had been filled (which would eliminate 1D rect), hence the exception for rect in that constraint. Without this second constraint, rows or columns appeared even for 1D rect encodings (e.g. multiple 1D rect encodings lined up). My thinking was that rect should be exempt from the first constraint, but all encodings should match the second constraint. It is a little redundant, hence my comment about splitting into just 'nonpositional' (maybe there is a better word for this that just encompasses color / size but excludes 'row' / 'column') and 'facet'.